### PR TITLE
feature: add possibility to use net.naming-scheme

### PR DESCRIPTION
**Pull request by MichalHe**
Leapp writes .link files to prevent interfaces being renamed
after booting to post-upgrade system. This patch adds a less
error-prone approach that uses net.naming-scheme kernel param.
The naming-scheme tells udev what hardware properties to use
when composing a device name. Moreover, possible values of this
parameter are coarse-grained "profiles", that tell udev to
behave as if it did on RHEL8.0.

The functionality is enabled by setting LEAPP_USE_NET_NAMING_SCHEME
environmental variable to 1. If the feature is enabled, the .link
file generation is disabled. A kernel parameter `net.naming-scheme=`
is added to the upgrade boot entry and the post-upgrade entry.
The value of the parameter will be `rhel-<source_major>.0`. Note
that the minor source version is *not used*. Using also source major
version instead of 0 causes the device names to change slightly,
so we use 0. Moreover, an extra RPM named `rhel-net-naming-sysattrs`
is installed to the target system and target userspace container.
The RPM provides definitions of the "profiles" for net.naming-scheme.

The feature is available only for 8>9 and higher. Attempting to
upgrade 7>8 with LEAPP_USE_NET_NAMING_SCHEME=1 will ignore
the value of LEAPP_USE_NET_NAMING_SCHEME.

Add a possibility to use the net.naming-scheme cmdline argument
to make immutable network interface names during the upgrade.
The feature can be used only for 8>9 upgrades and higher.
To enable the feature, use LEAPP_USE_NET_NAMING_SCHEME=1.

Jira-ref: RHEL-23473

### CODE REVIEW

**Comment by MichalHe** [comment]
I have tested the PR using a rhel8 system running on real HW with a NIC that would be renamed during the upgrade. The solutions installed the `rhel-net-naming-sysattrs` package to the target system, the NIC names were consistent, and the post-upgrade system was booted with `net.naming-scheme=rhel-8.10`.

**Review by matejmatuska — CHANGES_REQUESTED** [review]
Code-wise seems okay, just some minor comments.

I know RHEL7 is EOL, but maybe you could check if the `LEAPP_USE_NET_NAMING_SCHEME` envar is set for 7->8 upgrade. I would expect at least a warning message in the logs, since we can't and won't do what the user explicitly wants.

I haven't tested the code yet.

File: repos/system_upgrade/el8toel9/actors/emit_net_naming_scheme/tests/test_emit_net_naming_scheme.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        assert userspace_tasks.install_rpms == [emit_net_naming_lib.NET_NAMING_SYSATTRS_RPM_NAME]
+
+        rpm_tasks = ensure_one_msg_of_type_produced(produced_messages, RpmTransactionTasks)
+        assert rpm_tasks.to_install == [emit_net_naming_lib.NET_NAMING_SYSATTRS_RPM_NAME]
+    else:
+        assert not api.produce.called
+        return
+
+    if is_current_cmdline_compatible:
+        pass
```
[the remark this answers is not in the document]
Comment (MichalHe) [reply repos/system_upgrade/el8toel9/actors/emit_net_naming_scheme/tests/test_emit_net_naming_scheme.py:85]: Suggestion applied, thank you.

File: repos/system_upgrade/common/actors/addupgradebootentry/tests/unit_test_addupgradebootentry.py [hunk 2]
```diff
@@ -132,8 +132,13 @@ def get_boot_file_paths_mocked():
     assert addupgradebootentry.run.args[2] == run_args_add + ['-c', CONFIGS[0]]
     assert addupgradebootentry.run.args[3] == run_args_add + ['-c', CONFIGS[1]]
     assert api.produce.model_instances == [
+<<<<<<< HEAD
         TargetKernelCmdlineArgTasks(to_remove=[KernelCmdlineArg(key='debug')]),
         TargetKernelCmdlineArgTasks(to_remove=[KernelCmdlineArg(key='enforcing', value='0')]),
+=======
+        LateTargetKernelCmdlineArgTasks(to_remove=[KernelCmdlineArg(key='debug'),
+                                                   KernelCmdlineArg(key='enforcing', value='0')])
+>>>>>>> afb7fe51 (late_kernel_cmdline: break dependency loop)
```
[the remark this answers is not in the document]
Comment (MichalHe) [reply repos/system_upgrade/common/actors/addupgradebootentry/tests/unit_test_addupgradebootentry.py:141]: Cool, silly me. Should be fixed.

File: repos/system_upgrade/common/actors/addupgradebootentry/libraries/addupgradebootentry.py [hunk 3]
```diff
@@ -113,18 +118,65 @@ def figure_out_commands_needed_to_add_entry(kernel_path, initramfs_path, args_to
...
 def add_boot_entry(configs=None):
     kernel_dst_path, initram_dst_path = get_boot_file_paths()
+
     _remove_old_upgrade_boot_entry(kernel_dst_path, configs=configs)
 
     livemode_enabled = next(api.consume(LiveImagePreparationInfo), None) is not None
 
-    cmdline_args = collect_boot_args(livemode_enabled)
+    # Boot (desired/unwanted) for the upgrade kernel - we have to keep them separate as in order to guarantee
+    # the removal of undesired args we must call grubby in a separate call
```
[the remark this answers is not in the document]
Comment (MichalHe) [reply repos/system_upgrade/common/actors/addupgradebootentry/libraries/addupgradebootentry.py:172]: I have reworded it

File: repos/system_upgrade/common/actors/addupgradebootentry/libraries/addupgradebootentry.py [hunk 4]
```diff
@@ -113,18 +118,65 @@ def figure_out_commands_needed_to_add_entry(kernel_path, initramfs_path, args_to
...
+    args_not_present_on_target_kernel = added_upgrade_kernel_args - actual_kernel_args - target_args_to_add
+
+    # We remove only what we've added and what will not be already removed by someone else.
+    args_to_remove = [KernelCmdlineArg(key=arg[0], value=arg[1]) for arg in args_not_present_on_target_kernel]
+
+    if args_to_remove:
+        msg = ('Following upgrade kernel args were added, but they should not be present '
+               'on target cmdline: `%s`, requesting removal.')
+        api.current_logger().info(msg, args_not_present_on_target_kernel)
+        args_sorted = sorted(args_to_remove, key=lambda arg: arg.key)
```
[the remark this answers is not in the document]
Comment (MichalHe) [reply repos/system_upgrade/common/actors/addupgradebootentry/libraries/addupgradebootentry.py:160]: Of course - for deterministic testing

File: repos/system_upgrade/el8toel9/actors/emit_net_naming_scheme/libraries/emit_net_naming.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+                allows_predictable_names = True
+        if param.key == 'net.naming-scheme':
+            # We assume that the kernel cmdline does not contain invalid entries, namely,
+            # that the net.naming-scheme refers to a valid scheme.
+            already_has_a_net_naming_scheme = True
+
+    is_compatible = allows_predictable_names and not already_has_a_net_naming_scheme
+
+    msg = 'Should net.naming-scheme be added to kernel cmdline: %s'
+    api.current_logger().info(msg, 'yes' if is_compatible else 'no')
```
[the remark this answers is not in the document]
Comment (MichalHe) [reply repos/system_upgrade/el8toel9/actors/emit_net_naming_scheme/libraries/emit_net_naming.py:38]: Oky :)

File: repos/system_upgrade/common/actors/addupgradebootentry/libraries/addupgradebootentry.py [hunk 6]
```diff
@@ -113,18 +118,65 @@ def figure_out_commands_needed_to_add_entry(kernel_path, initramfs_path, args_to
...
+
+
+def fmt_kernel_args(args):
+    def fmt_arg(arg):
+        if arg[1]:
+            return '{0}={1}'.format(*arg)
+        return arg[0]
+
+    args_str = ' '.join(fmt_arg(arg) for arg in sorted(args, key=lambda arg: arg[0]))
+    return args_str
```
[the remark this answers is not in the document]
Comment (MichalHe) [reply repos/system_upgrade/common/actors/addupgradebootentry/libraries/addupgradebootentry.py]: removed

**Comment by matejmatuska** [comment]
Tested, now works as expected.

PLease rebase and resolved conflicts.

**Comment by MichalHe** [comment]
@matejmatuska rebased

**Review by matejmatuska — APPROVED** [review]
For now this is only going to be enabled on 8->9.

The `persistentnamesconfig` actor is now properly skipped on paths other than 8->9 and `emit_net_naming_scheme` only emits on 8->9.
I tested it and it works as expected. LGTM

**Pull request merged**